### PR TITLE
Allow AD groups as Project users

### PR DESCRIPTION
**Pull request by dmettem**
Fix proposal is to deprecate the existing arguments for administrators, members, and viewers
and introduce new arguments to allow both email addresses and AD groups.

Signed-off-by: Deepak Mettem <dmettem@vmware.com>

### CODE REVIEW

File: vra/resource_project.go [hunk 1]
```diff
@@ -20,11 +20,13 @@ func resourceProject() *schema.Resource {
 			"administrators": {
 				Type:        schema.TypeSet,
 				Optional:    true,
+				Deprecated:  "To specify the type of principal, please refer administrator.",
 				Description: "List of administrator users associated with the project. Only administrators can manage project's configuration.",
 				Elem: &schema.Schema{
 					Type: schema.TypeString,
 				},
 			},
+			"administrator": userSchema(),
```
Comment (Prativa20) [inline vra/resource_project.go:29]: We should have a description for this property

File: vra/resource_project.go [hunk 2]
```diff
@@ -56,11 +58,13 @@ func resourceProject() *schema.Resource {
 			"members": {
 				Type:        schema.TypeSet,
 				Optional:    true,
+				Deprecated:  "To specify the type of principal, please refer member.",
 				Description: "List of member users associated with the project.",
 				Elem: &schema.Schema{
 					Type: schema.TypeString,
 				},
 			},
+			"member": userSchema(),
```
Comment (Prativa20) [inline vra/resource_project.go:67]: same as above

File: vra/resource_project.go [hunk 3]
```diff
@@ -80,11 +84,13 @@ func resourceProject() *schema.Resource {
 			"viewers": {
 				Type:        schema.TypeSet,
 				Optional:    true,
+				Deprecated:  "To specify the type of principal, please refer viewer.",
 				Description: "List of viewer users associated with the project.",
 				Elem: &schema.Schema{
 					Type: schema.TypeString,
 				},
 			},
+			"viewer": userSchema(),
```
Comment (Prativa20) [inline vra/resource_project.go:93]: same as above

File: examples/project/main.tf [hunk 4]
```diff
@@ -25,10 +25,40 @@ resource "vra_project" "this" {
 
   administrators = ["jason@vra.local"]
 
+  administrator {
+    email = "jason@vra.local"
+    type = "user"
+  }
+
+  administrator {
```
Comment (Prativa20) [inline examples/project/main.tf:33]: should we make this administrator_list instead?

File: examples/project/main.tf [hunk 5]
```diff
@@ -25,10 +25,40 @@ resource "vra_project" "this" {
...
+    email = "jason-group@vra.local"
+    type = "group"
+  }
+
   members = ["tony@vra.local"]
 
+  member {
```
Comment (Prativa20) [inline examples/project/main.tf:40]: member_list?

File: examples/project/main.tf [hunk 6]
```diff
@@ -25,10 +25,40 @@ resource "vra_project" "this" {
...
+  }
+
+  member {
+    email = "tony-group@vra.local"
+    type = "group"
+  }
+
   viewers = ["shauna@vra.local"]
 
+  viewer {
```
Comment (Prativa20) [inline examples/project/main.tf:52]: viewer_list?

**Review by Prativa20 — COMMENTED** [review]
A few minor comments. We should also update the acceptance tests and add this new property if needed

File: vra/resource_project.go (hunk 1 above)
Comment (dondon3345678) [reply vra/resource_project.go]: Rename to `administrator_roles`

File: vra/resource_project.go (hunk 2 above)
Comment (dondon3345678) [reply vra/resource_project.go]: Rename to `member_roles`

File: examples/project/main.tf (hunk 4 above)
Comment (dondon3345678) [reply examples/project/main.tf]: Same as above

File: examples/project/main.tf (hunk 5 above)
Comment (dondon3345678) [reply examples/project/main.tf]: Same as above

File: examples/project/main.tf (hunk 6 above)
Comment (dondon3345678) [reply examples/project/main.tf]: Same as above

File: vra/resource_project.go [hunk 7]
```diff
@@ -20,11 +20,13 @@ func resourceProject() *schema.Resource {
 			"administrators": {
 				Type:        schema.TypeSet,
 				Optional:    true,
+				Deprecated:  "To specify the type of principal, please refer administrator.",
```
Comment (Prativa20) [inline vra/resource_project.go:23]: please refer administrator_roles*

File: vra/resource_project.go [hunk 8]
```diff
@@ -56,11 +58,13 @@ func resourceProject() *schema.Resource {
 			"members": {
 				Type:        schema.TypeSet,
 				Optional:    true,
+				Deprecated:  "To specify the type of principal, please refer member.",
```
Comment (Prativa20) [inline vra/resource_project.go:61]: please refer member_roles*

File: vra/resource_project.go [hunk 9]
```diff
@@ -80,11 +84,13 @@ func resourceProject() *schema.Resource {
 			"viewers": {
 				Type:        schema.TypeSet,
 				Optional:    true,
+				Deprecated:  "To specify the type of principal, please refer viewer.",
```
Comment (Prativa20) [inline vra/resource_project.go:87]: please refer viewer_roles*

File: vra/resource_project.go [hunk 10]
```diff
@@ -179,16 +186,16 @@ func resourceProjectRead(d *schema.ResourceData, m interface{}) error {
...
 	d.Set("description", project.Description)
 	d.Set("machine_naming_template", project.MachineNamingTemplate)
-	d.Set("members", flattenUserList(project.Members))
+	d.Set("members", flattenUsers(project.Members))
 	d.Set("name", project.Name)
 	d.Set("operation_timeout", project.OperationTimeout)
 	d.Set("shared_resources", project.SharedResources)
-	d.Set("viewers", flattenUserList(project.Viewers))
+	d.Set("viewers", flattenUsers(project.Viewers))
 	d.Set("zone_assignments", flattenZoneAssignment(project.Zones))
```
Comment (Prativa20) [inline vra/resource_project.go:199]: We should set the administrator_roles, member_roles and viewer_roles in the read method to reflect in the state file

File: website/docs/r/vra_project.html.markdown [hunk 11]
```diff
@@ -29,12 +29,45 @@ resource "vra_project" "this" {
   
   shared_resources = false
 
+  # Deprecated
```
Comment (Prativa20) [inline website/docs/r/vra_project.html.markdown:32]: Can you also mention, "use administrator_roles instead"

File: vra/resource_project.go (hunk 7 above)
Comment (dondon3345678) [reply vra/resource_project.go:23]: Done

File: vra/resource_project.go (hunk 8 above)
Comment (dondon3345678) [reply vra/resource_project.go]: Done

File: vra/resource_project.go (hunk 9 above)
Comment (dondon3345678) [reply vra/resource_project.go]: Done

File: vra/resource_project.go (hunk 10 above)
Comment (dondon3345678) [reply vra/resource_project.go]: Done

File: website/docs/r/vra_project.html.markdown (hunk 11 above)
Comment (dondon3345678) [reply website/docs/r/vra_project.html.markdown]: Done

**Review by Prativa20 — APPROVED** [review]
Looks good!

**Pull request merged**